src/main/java/teammates/ui/webapi/action/GetStudentProfilePictureAction.java

```diff
@@ -1,25 +1,73 @@
 package teammates.ui.webapi.action;
 
+import org.apache.http.HttpStatus;
+
+import teammates.common.datatransfer.attributes.StudentAttributes;
+import teammates.common.datatransfer.attributes.StudentProfileAttributes;
+import teammates.common.exception.UnauthorizedAccessException;
 import teammates.common.util.Const;
 
 /**
  * Action: serves a profile picture that is stored in Google Cloud Storage.
  */
 public class GetStudentProfilePictureAction extends Action {
 
+    /** String indicating ACCESS is not given. */
+    public static final String UNAUTHORIZED_ACCESS = "You are not allowed to view this resource!";
+
+    /** String indicating student not found. */
+    public static final String STUDENT_NOT_FOUND = "No student found for given parameters";
+
+    /** String indicating profile picture not found. */
+    public static final String PROFILE_PIC_NOT_FOUND = "Student has no profile picture";
+
     @Override
     protected AuthType getMinAuthLevel() {
         return AuthType.LOGGED_IN;
     }
 
     @Override
     public void checkSpecificAccessControl() {
-        // Profile picture is available to everyone with the blob key
+        String courseId = getRequestParamValue(Const.ParamsNames.COURSE_ID);
+        String studentEmail = getRequestParamValue(Const.ParamsNames.STUDENT_EMAIL);
+
+        //viewing someone else's photo
+        if (studentEmail != null && courseId != null) {
+            StudentAttributes student = logic.getStudentForEmail(courseId, studentEmail);
+
+            if (student == null) {
+                throw new UnauthorizedAccessException(UNAUTHORIZED_ACCESS);
```

Comment:
This does not looks like `UnauthorizedAccessException` to me, should it be 404 instead?

Reply:
Yes it is technically a 404. 

However, this is under the checkAccessControl method. This method checks access (it doesnt deal with action logic). By convention I have been throwing unauthorized access here.

Reply:
Even in the `checkSpecificAccessControl` method?

That method is of `void` type and hence you cant return a `new JsonResult(..., HttpStatus.SC_NOT_FOUND)` as we normally do in our `Action`. As a result, for almost all of the actions here, I have returned unauthorized access if I cant find the student/instructor.